Refactor and unit test symex_assign_symbol

romainbrenguier opened the pull request:
~This is based on https://github.com/diffblue/cbmc/pull/4825 (first 10 commits).~ (merged and rebased)
The goal was to add unit test for symex_assign_symbol, to make that easier we make `symex_assign_rec` and the functions it calls static, and add a header `symex_assign.h` which corresponds to `symex_assign.cpp`.



<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Extract a static assign_non_struct_symbol function

This will make it easier to unit test this part of the program in
isolation of the rest of symex.

romainbrenguier committed:
Make symex_assign_from_struct static

This will make it easier to unit-test it in isolation from the rest of
the program.

romainbrenguier committed:
Move debug output to symex_assign

We move this debug call to the main function so that intermediary
functions don't need to access the log field, which makes it easier to
make them static. This also ensures there is one output per assignment.

romainbrenguier committed:
Make symex_assign_rec non-member

This requires making intermediary functions non-members as well.
This will make it easier to unit-test this part of symex in isolation
from the rest.

romainbrenguier committed:
Move goto_symext::symex_assign to goto_symex.cpp

symex_assign.cpp now only contains functions that are not members of
goto_symext.
This will make the separation between members and non-members clearer.

romainbrenguier committed:
Add a symex_assign.h header

This puts the definition of the useful functions from symex_assign.cpp
in a symex_assign.h. This makes navigating the code more natural.

romainbrenguier committed:
Add unit test for symex_assign_symbol

This shows the behaviour of the function on a simple example.

romainbrenguier committed:
Test ssa/original_full_lhs in symex_assign_symbol

Test what these values are expected to be.

romainbrenguier committed:
Test symex_assign with constant propagation

Test that constant values get propagated and that this does not
interfer with a second assignment.

romainbrenguier committed:
Document symex_assign_symbol

Clarifies the role of this function

romainbrenguier committed:
Define symex_assignt class

This groups together arguments that do not change between recursive
calls to the symex_assign_* functions.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1afc94b2fc10fb23e4d9b0744f2126c4ccec3cf1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116617890

smowton commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -407,13 +421,26 @@ void goto_symext::symex_assign_from_struct(
-      comp_rhs.second,
-      guard,
-      assignment_type);
+    if(comp_rhs.second.id() == ID_struct)
```
This is duplicated from `symex_assign_symbol`, extract that too?

smowton commented on unit/goto-symex/symex_assign.cpp:
```diff
@@ -0,0 +1,124 @@
+          const symbol_exprt rhs_symbol = to_symbol_expr(rhs_if->false_case());
+          REQUIRE(rhs_symbol.get_identifier() == "foo!0#0");
+        }
+
```
Extra whitespace

smowton commented on unit/goto-symex/symex_assign.cpp:
```diff
@@ -0,0 +1,124 @@
+        SSA_stept step = target.SSA_steps.back();
+        REQUIRE(step.type == goto_trace_stept::typet::ASSIGNMENT);
+        REQUIRE(step.assignment_type == symex_targett::assignment_typet::STATE);
+        REQUIRE(step.cond_expr.id() == ID_equal);
```
Seems redundant with the check below

smowton commented on unit/goto-symex/symex_assign.cpp:
```diff
@@ -129,4 +126,73 @@ SCENARIO(
+    "activated")
+  {
+    exprt::operandst guard;
+    null_message_handlert log;
```
There's already a null message handler in https://github.com/diffblue/cbmc/blob/develop/unit/testing-utils/message.h

smowton commented on unit/goto-symex/symex_assign.cpp:
```diff
@@ -129,4 +126,73 @@ SCENARIO(
+        SSA_stept step = target_equation.SSA_steps.back();
+        REQUIRE(step.guard == true_exprt{});
+      }
+      THEN("The propagation map contains an entry for foo")
```
The propagation map maps "foo" -> 5721

smowton commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -13,8 +13,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/format_expr.h>
-#include <util/simplify_expr.h>
 #include <util/pointer_offset_size.h>
+#include <util/simplify_expr.h>
```
Don't use en-masse clang-format commits, instead format-and-squash at each commit

smowton reviewed (commented):
I agree that pulling the assign-related mutually recursive functions into their own class, which only carries around the information necessary for executing an assignment, makes them easier to test, but turning that class into a parameter pack named `info` and making all the functions static (effectively just making all that class' functions public and substituting implicit `this` for `info` everywhere) seems like a lot of churn for marginal benefit to me. What would this look like if we just created `goto_symex_assignt` and made the symex-assign-* functions members of that class instead of `goto_symext`?

Also since assign-symbol is a very hot function I'm a little worried about anonymous function used to initialise the info struct, though perhaps the compiler is smart enough for that.

romainbrenguier commented:
@smowton 
> I agree that pulling the assign-related mutually recursive functions into their own class, which only carries around the information necessary for executing an assignment, makes them easier to test, but turning that class into a parameter pack named info and making all the functions static (effectively just making all that class' functions public and substituting implicit this for info everywhere) seems like a lot of churn for marginal benefit to me. What would this look like if we just created goto_symex_assignt and made the symex-assign-* functions members of that class instead of goto_symext?

That would be basically the same thing, but if the `goto_symex_assignt` class would be defined in the header then all its method need to be declared there, which I find distracting when navigating header file as the public methods get lost among the private ones. On the other hand, a class would make the syntax in the body of the functions a bit nicer. So I don't mind it being done the other way but my preference goes to having the nicer interface rather than nicer field access in the code.

romainbrenguier committed:
Use symex_assign_symbol in symex_assign_from_struct

Deduplicate code that was similar an symex_assign_from_struct and
symex_assign_symbol.

smowton commented:
Guess we need a tie-breaker from @tautschnig @peterschrammel @kroening, but my preference goes to a class, both to reduce churn and to deviate less from the codebase norms

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 667c1457ab6cdb402104cf95f7abec878b262a18).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116642140
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f0e5c845ed20f8ad13599052742f82fc59338c06).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116645015

peterschrammel commented:
I vote for 'class'.

romainbrenguier commented:
@peterschrammel @smowton I replaced the info struct by a class

smowton commented:
Keep the `assign_` prefixes on the function names (function names should contain a verb), otherwise looks good

romainbrenguier committed:
Shorten method names in symex_assignt

symex is already included in the class name and does not need to be
repeated.

romainbrenguier commented:
@smowton 
> Keep the assign_ prefixes on the function names (function names should contain a verb), otherwise looks good

Done, but I disagree with the general statement that function names should be verbs, there are plenty of examples of the opposite in the standard library (for instance std::vector has size, max_size, capacity, empty, at, front, back, data), and for a functor like here, the verb is already in the class/object name.

smowton commented:
Sure ok, but those are properties or accessors; in this case we have an imperative action. Without the actual action specifier you get code like `symbol(s);`, which suggests to me "make a symbol out of s" or "get the symbol for s", rather than "assign to symbol s", which `assign_symbol` makes clear.

smowton reviewed (approved):
There's likely some commits to squash, otherwise lgtm

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 37321a7762715a6017450ffd8a94bf88748d6afb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116808254

peterschrammel commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -20,3 +23,59 @@ void goto_symext::do_simplify(exprt &expr)
+    if(
+      statement == ID_cpp_new || statement == ID_cpp_new_array ||
+      statement == ID_java_new_array_data)
+      symex_cpp_new(state, lhs, side_effect_expr);
```
⛏  braces after multi-line condition

peterschrammel commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -20,3 +23,59 @@ void goto_symext::do_simplify(exprt &expr)
+      lhs.id() == ID_symbol &&
+      id2string(to_symbol_expr(lhs).get_identifier()).find("#return_value!") !=
+        std::string::npos)
+      assignment_type = symex_targett::assignment_typet::HIDDEN;
```
⛏ braces after multi-line condition

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d929d3443f3a8a636e930a4c57dde22be7dddcab).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116823785
